File: core/src/main/java/com/vzome/core/kinds/PolygonFieldApplication.java
package com.vzome.core.kinds;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import com.vzome.api.Tool;
import com.vzome.api.Tool.Factory;
import com.vzome.core.algebra.AlgebraicNumber;
import com.vzome.core.algebra.PolygonField;
import com.vzome.core.commands.Command;
import com.vzome.core.commands.CommandAxialSymmetry;
import com.vzome.core.commands.CommandUniformH4Polytope;
import com.vzome.core.editor.SymmetryPerspective;
import com.vzome.core.editor.ToolsModel;
import com.vzome.core.math.symmetry.AntiprismSymmetry;
import com.vzome.core.math.symmetry.IcosahedralSymmetry;
import com.vzome.core.math.symmetry.QuaternionicSymmetry;
import com.vzome.core.math.symmetry.WythoffConstruction.Listener;
import com.vzome.core.tools.AxialStretchTool;
import com.vzome.core.tools.AxialSymmetryToolFactory;
import com.vzome.core.tools.IcosahedralToolFactory;
import com.vzome.core.tools.LinearMapTool;
import com.vzome.core.tools.MirrorTool;
import com.vzome.core.tools.RotationTool;
import com.vzome.core.tools.ScalingTool;
import com.vzome.core.tools.SymmetryTool;
import com.vzome.core.tools.TranslationTool;
import com.vzome.core.viewing.AbstractShapes;
import com.vzome.core.viewing.AntiprismShapes;
import com.vzome.core.viewing.OctahedralShapes;

/**
 * Everything here is stateless, or at worst, a cache (like Shapes).
 * An instance of this can be shared by many DocumentModels.
 * This is why it does not have tool factories, though it does
 * dictate what tool factories will be present.
 *
 * @author David Hall
 *
 */
public class PolygonFieldApplication extends DefaultFieldApplication
{
    public PolygonFieldApplication(int polygonSides)
    {
        super( new PolygonField(polygonSides) );
        symmetryPerspectives.add( new AntiprismSymmetryPerspective());    
        
        if(polygonSides == 5) { // TODO: eventually use getField().getGoldenRatio() != null here 
            // when we eventually generalize icosa symm and strut rendering to work with any 5N-gon.
            // Until then, enable icosa symm only for polygon(5) since it only works in that case for now.
            icosahedralPerspective = new IcosahedralSymmetryPerspective(getField());
            symmetryPerspectives.add(icosahedralPerspective);
            H4 = new QuaternionicSymmetry( "H_4", "com/vzome/core/math/symmetry/H4roots.vef", getField() );
        } else {
            icosahedralPerspective = null;
            H4 = null;
        }
        symmetryPerspectives.add(super.getDefaultSymmetryPerspective());
    }

    @Override
    public PolygonField getField() {
        // This cast to PolygonField is safe 
        // because a PolygonField is used in the c'tor
        return (PolygonField) super.getField();
    }
    
    private class AntiprismSymmetryPerspective extends AbstractSymmetryPerspective 
    {
        AntiprismSymmetryPerspective() {
            super(new AntiprismSymmetry(getField()).createStandardOrbits( "blue" ));
            AbstractShapes thinAntiprismShapes = new AntiprismShapes( "thin", "thin antiprism", getSymmetry() );
            AbstractShapes antiprismShapes = new AntiprismShapes( "antiprism", "antiprism", getSymmetry() );
            // TODO: Trapezohedron connectors (dual of Antiprism)
            AbstractShapes octahedralShapes = new OctahedralShapes( "octahedral", "octahedral", symmetry );
            
            // this is the order they will be shown on the dialog
            setDefaultGeometry(thinAntiprismShapes);
            addShapes(antiprismShapes);
            addShapes(octahedralShapes);
        }
        
        @Override
        public AntiprismSymmetry getSymmetry() {
            // This cast to AntiprismSymmetry is safe 
            // because an AntiprismSymmetry is used in the c'tor
            return (AntiprismSymmetry) super.getSymmetry();
        }
        
        @Override
        public List<Tool.Factory> createToolFactories(Tool.Kind kind, ToolsModel tools) {
            List<Tool.Factory> result = new ArrayList<>();
            switch (kind) {

            case SYMMETRY:
                result.add(new SymmetryTool.Factory(tools, this.symmetry));
                result.add(new MirrorTool.Factory(tools));
                result.add(new AxialSymmetryToolFactory(tools, this.symmetry));
                break;

            case TRANSFORM:
                result.add(new ScalingTool.Factory(tools, this.symmetry));
                result.add(new RotationTool.Factory(tools, this.symmetry));
                result.add(new TranslationTool.Factory(tools));
                break;

            case LINEAR_MAP:
                result.add(new LinearMapTool.Factory(tools, this.symmetry, false));
                break;

            default:
                break;
            }
            return result;
        }

        @Override
        public List<Tool> predefineTools(Tool.Kind kind, ToolsModel tools) {
            List<Tool> result = new ArrayList<>();
            switch (kind) {

            case SYMMETRY:
                result.add(new SymmetryTool.Factory(tools, this.symmetry).createPredefinedTool("polygonal antiprism around origin"));
                result.add(new MirrorTool.Factory(tools).createPredefinedTool("reflection through XY plane"));
                result.add(new AxialSymmetryToolFactory(tools, this.symmetry, true).createPredefinedTool("symmetry around red through origin"));
                break;

            case TRANSFORM:
                result.add(new ScalingTool.Factory(tools, this.symmetry).createPredefinedTool("scale down"));
                result.add(new ScalingTool.Factory(tools, this.symmetry).createPredefinedTool("scale up"));
                result.add(new RotationTool.Factory(tools, this.symmetry, true).createPredefinedTool("rotate around red through origin"));
                break;

            default:
                break;
            }
            return result;
        }

        private final Command axialsymm = new CommandAxialSymmetry( symmetry );

        @Override
        public Command getLegacyCommand(String action) {
            switch (action) {
            case "axialsymm":
                return axialsymm;
            default:
                return super.getLegacyCommand(action);
            }
        }

        @Override
        public String getModelResourcePath() {
            // getTrackballModelStream() dynamically modifies this template file
            return "org/vorthmann/zome/app/antiprism-trackball-template.vZome";
        }
    };

    ArrayList<SymmetryPerspective> symmetryPerspectives = new ArrayList<>();
    
    @Override
    public Collection<SymmetryPerspective> getSymmetryPerspectives() {
        return symmetryPerspectives;
    }

    private final IcosahedralSymmetryPerspective icosahedralPerspective;

    @Override
    public SymmetryPerspective getDefaultSymmetryPerspective() {
        return (icosahedralPerspective == null) ? symmetryPerspectives.get(0) : icosahedralPerspective;
    }

    @Override
    public SymmetryPerspective getSymmetryPerspective(String symmName) {
        for(SymmetryPerspective sp : symmetryPerspectives) {
            if(sp.getName().equals(symmName)) {
                return sp;
            }
        }
        return super.getSymmetryPerspective(symmName);
    }

    @Override
    public void registerToolFactories( Map<String, Factory> toolFactories, ToolsModel tools )
    {
        // register the default tool factories
        super.registerToolFactories( toolFactories, tools );
        
        // add any tools that are unique for this field
        if(icosahedralPerspective != null) {
            IcosahedralSymmetry symm = icosahedralPerspective .getSymmetry();
            // symm matters for this one, since it is final in the tool
            toolFactories .put( "AxialStretchTool", new AxialStretchTool.Factory( tools, symm, false, false, false ) );
            // this one has to replace the same-named factory in the base class
            toolFactories .put( "SymmetryTool", new IcosahedralToolFactory( tools, symm ) );
        }        
    }
    
    private final QuaternionicSymmetry H4;
    
    @Override
    public QuaternionicSymmetry getQuaternionSymmetry( String name )
    {
        switch ( name ) {

        case "H_4":
            return this .H4;

        default:
            return null;
        }
    }

    private CommandUniformH4Polytope h4Builder = null;
    
    @Override
    public void constructPolytope( String groupName, int index, int edgesToRender, AlgebraicNumber[] edgeScales, Listener listener )
    {
        switch ( groupName ) {

        case "H4":
            if ( this .h4Builder == null ) {
                QuaternionicSymmetry qsymm = new QuaternionicSymmetry( "H_4", "com/vzome/core/math/symmetry/H4roots.vef", this .getField() );
                this .h4Builder = new CommandUniformH4Polytope( this .getField(), qsymm, 0 );
            }
            this .h4Builder .generate( index, edgesToRender, edgeScales, listener );
            break;

        default:
            super .constructPolytope( groupName, index, edgesToRender, edgeScales, listener );
            break;
        }
    }

}

File: online/src/worker/legacy/from-java/com/vzome/core/kinds/PolygonFieldApplication.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { Tool } from "../../api/Tool.js";
import { AlgebraicNumber } from "../algebra/AlgebraicNumber.js";
import { PolygonField } from "../algebra/PolygonField.js";
import { Command } from "../commands/Command.js";
import { CommandAxialSymmetry } from "../commands/CommandAxialSymmetry.js";
import { CommandUniformH4Polytope } from "../commands/CommandUniformH4Polytope.js";
import { SymmetryPerspective } from "../editor/SymmetryPerspective.js";
import { ToolsModel } from "../editor/ToolsModel.js";
import { AbstractSymmetryPerspective } from "./AbstractSymmetryPerspective.js";
import { DefaultFieldApplication } from "./DefaultFieldApplication.js";
import { IcosahedralSymmetryPerspective } from "./IcosahedralSymmetryPerspective.js";
import { AntiprismSymmetry } from "../math/symmetry/AntiprismSymmetry.js";
import { IcosahedralSymmetry } from "../math/symmetry/IcosahedralSymmetry.js";
import { QuaternionicSymmetry } from "../math/symmetry/QuaternionicSymmetry.js";
import { WythoffConstruction } from "../math/symmetry/WythoffConstruction.js";
import { AxialStretchTool } from "../tools/AxialStretchTool.js";
import { AxialSymmetryToolFactory } from "../tools/AxialSymmetryToolFactory.js";
import { IcosahedralToolFactory } from "../tools/IcosahedralToolFactory.js";
import { InversionToolFactory } from "../tools/InversionToolFactory.js";
import { LineReflectionToolFactory } from "../tools/LineReflectionToolFactory.js";
import { LinearMapToolFactory } from "../tools/LinearMapToolFactory.js";
import { MirrorToolFactory } from "../tools/MirrorToolFactory.js";
import { ProjectionToolFactory } from "../tools/ProjectionToolFactory.js";
import { RotationToolFactory } from "../tools/RotationToolFactory.js";
import { ScalingToolFactory } from "../tools/ScalingToolFactory.js";
import { SymmetryToolFactory } from "../tools/SymmetryToolFactory.js";
import { TranslationToolFactory } from "../tools/TranslationToolFactory.js";
import { AbstractShapes } from "../viewing/AbstractShapes.js";
import { AntiprismShapes } from "../viewing/AntiprismShapes.js";
import { OctahedralShapes } from "../viewing/OctahedralShapes.js";

/**
 * Everything here is stateless, or at worst, a cache (like Shapes).
 * An instance of this can be shared by many DocumentModels.
 * This is why it does not have tool factories, though it does
 * dictate what tool factories will be present.
 * 
 * @author David Hall
 * @param {PolygonField} field
 * @class
 * @extends DefaultFieldApplication
 */
export class PolygonFieldApplication extends DefaultFieldApplication {
    public constructor(field: PolygonField) {
        super(field);
        this.symmetryPerspectives = <any>(new java.util.ArrayList<any>());
        if (this.icosahedralPerspective === undefined) { this.icosahedralPerspective = null; }
        if (this.H4 === undefined) { this.H4 = null; }
        this.h4Builder = null;
        this.symmetryPerspectives.add(new PolygonFieldApplication.AntiprismSymmetryPerspective(this));
        if (field.polygonSides() === 5){
            this.icosahedralPerspective = new IcosahedralSymmetryPerspective(this.getField());
            this.symmetryPerspectives.add(this.icosahedralPerspective);
            this.H4 = new QuaternionicSymmetry("H_4", "com/vzome/core/math/symmetry/H4roots.vef", this.getField());
        } else {
            this.icosahedralPerspective = null;
            this.H4 = null;
        }
        this.symmetryPerspectives.add(super.getDefaultSymmetryPerspective());
    }

    /**
     * 
     * @return {PolygonField}
     */
    public getField(): PolygonField {
        return <PolygonField><any>super.getField();
    }

    symmetryPerspectives: java.util.ArrayList<SymmetryPerspective>;

    /**
     * 
     * @return {*}
     */
    public getSymmetryPerspectives(): java.util.Collection<SymmetryPerspective> {
        return this.symmetryPerspectives;
    }

    /*private*/ icosahedralPerspective: IcosahedralSymmetryPerspective;

    /**
     * 
     * @return {*}
     */
    public getDefaultSymmetryPerspective(): SymmetryPerspective {
        return (this.icosahedralPerspective == null) ? this.symmetryPerspectives.get(0) : this.icosahedralPerspective;
    }

    /**
     * 
     * @param {string} symmName
     * @return {*}
     */
    public getSymmetryPerspective(symmName: string): SymmetryPerspective {
        for(let index=this.symmetryPerspectives.iterator();index.hasNext();) {
            let sp = index.next();
            {
                if (sp.getName() === symmName){
                    return sp;
                }
            }
        }
        return super.getSymmetryPerspective(symmName);
    }

    /**
     * 
     * @param {*} toolFactories
     * @param {ToolsModel} tools
     */
    public registerToolFactories(toolFactories: java.util.Map<string, Tool.Factory>, tools: ToolsModel) {
        super.registerToolFactories(toolFactories, tools);
        if (this.icosahedralPerspective != null){
            const symm: IcosahedralSymmetry = this.icosahedralPerspective.getSymmetry();
            toolFactories.put("AxialStretchTool", new AxialStretchTool.Factory(tools, symm, false, false, false));
            toolFactories.put("SymmetryTool", new IcosahedralToolFactory(tools, symm));
        }
    }

    /*private*/ H4: QuaternionicSymmetry;

    /**
     * 
     * @param {string} name
     * @return {QuaternionicSymmetry}
     */
    public getQuaternionSymmetry(name: string): QuaternionicSymmetry {
        switch((name)) {
        case "H_4":
            return this.H4;
        default:
            return null;
        }
    }

    /*private*/ h4Builder: CommandUniformH4Polytope;

    /**
     * 
     * @param {string} groupName
     * @param {number} index
     * @param {number} edgesToRender
     * @param {AlgebraicNumber[]} edgeScales
     * @param {*} listener
     */
    public constructPolytope(groupName: string, index: number, edgesToRender: number, edgeScales: AlgebraicNumber[], listener: WythoffConstruction.Listener) {
        switch((groupName)) {
        case "H4":
            if (this.h4Builder == null){
                const qsymm: QuaternionicSymmetry = new QuaternionicSymmetry("H_4", "com/vzome/core/math/symmetry/H4roots.vef", this.getField());
                this.h4Builder = new CommandUniformH4Polytope(this.getField(), qsymm, 0);
            }
            this.h4Builder.generate(index, edgesToRender, edgeScales, listener);
            break;
        default:
            super.constructPolytope(groupName, index, edgesToRender, edgeScales, listener);
            break;
        }
    }
}
PolygonFieldApplication["__class"] = "com.vzome.core.kinds.PolygonFieldApplication";
PolygonFieldApplication["__interfaces"] = ["com.vzome.core.math.symmetry.Symmetries4D","com.vzome.core.editor.FieldApplication"];



export namespace PolygonFieldApplication {

    export class AntiprismSymmetryPerspective extends AbstractSymmetryPerspective {
        public __parent: any;
        constructor(__parent: any) {
            super(new AntiprismSymmetry(__parent.getField()).createStandardOrbits("blue"));
            this.__parent = __parent;
            this.axialsymm = new CommandAxialSymmetry(this.symmetry);
            const thinAntiprismShapes: AbstractShapes = new AntiprismShapes("thin", "thin antiprism", this.getSymmetry());
            const antiprismShapes: AbstractShapes = new AntiprismShapes("antiprism", "antiprism", this.getSymmetry());
            const octahedralShapes: AbstractShapes = new OctahedralShapes("octahedral", "octahedral", this.symmetry);
            this.setDefaultGeometry(thinAntiprismShapes);
            this.addShapes(antiprismShapes);
            this.addShapes(octahedralShapes);
        }

        /**
         * 
         * @return {string}
         */
        public getLabel(): string {
            return "antiprism";
        }

        /**
         * 
         * @return {AntiprismSymmetry}
         */
        public getSymmetry(): AntiprismSymmetry {
            return <AntiprismSymmetry><any>super.getSymmetry();
        }

        /**
         * 
         * @param {Tool.Kind} kind
         * @param {ToolsModel} tools
         * @return {*}
         */
        public createToolFactories(kind: Tool.Kind, tools: ToolsModel): java.util.List<Tool.Factory> {
            const isTrivial: boolean = this.symmetry.isTrivial();
            const result: java.util.List<Tool.Factory> = <any>(new java.util.ArrayList<any>());
            switch((kind)) {
            case Tool.Kind.SYMMETRY:
                result.add(new SymmetryToolFactory(tools, this.symmetry));
                if (isTrivial){
                    result.add(new InversionToolFactory(tools));
                }
                result.add(new LineReflectionToolFactory(tools));
                result.add(new MirrorToolFactory(tools));
                result.add(new AxialSymmetryToolFactory(tools, this.symmetry));
                break;
            case Tool.Kind.TRANSFORM:
                result.add(new ScalingToolFactory(tools, this.symmetry));
                result.add(new RotationToolFactory(tools, this.symmetry));
                result.add(new TranslationToolFactory(tools));
                if (isTrivial){
                    result.add(new ProjectionToolFactory(tools));
                }
                break;
            case Tool.Kind.LINEAR_MAP:
                result.add(new LinearMapToolFactory(tools, this.symmetry, false));
                break;
            default:
                break;
            }
            return result;
        }

        /**
         * 
         * @param {Tool.Kind} kind
         * @param {ToolsModel} tools
         * @return {*}
         */
        public predefineTools(kind: Tool.Kind, tools: ToolsModel): java.util.List<Tool> {
            const result: java.util.List<Tool> = <any>(new java.util.ArrayList<any>());
            switch((kind)) {
            case Tool.Kind.SYMMETRY:
                result.add(new SymmetryToolFactory(tools, this.symmetry).createPredefinedTool("polygonal antiprism around origin"));
                result.add(new MirrorToolFactory(tools).createPredefinedTool("reflection through XY plane"));
                result.add(new AxialSymmetryToolFactory(tools, this.symmetry, true).createPredefinedTool("symmetry around red through origin"));
                break;
            case Tool.Kind.TRANSFORM:
                result.add(new ScalingToolFactory(tools, this.symmetry).createPredefinedTool("scale down"));
                result.add(new ScalingToolFactory(tools, this.symmetry).createPredefinedTool("scale up"));
                result.add(new RotationToolFactory(tools, this.symmetry, true).createPredefinedTool("rotate around red through origin"));
                break;
            default:
                break;
            }
            return result;
        }

        axialsymm: Command;

        /**
         * 
         * @param {string} action
         * @return {*}
         */
        public getLegacyCommand(action: string): Command {
            switch((action)) {
            case "axialsymm":
                return this.axialsymm;
            default:
                return super.getLegacyCommand(action);
            }
        }

        /**
         * 
         * @return {string}
         */
        public getModelResourcePath(): string {
            return "org/vorthmann/zome/app/antiprism-trackball-template.vZome";
        }
    }
    AntiprismSymmetryPerspective["__class"] = "com.vzome.core.kinds.PolygonFieldApplication.AntiprismSymmetryPerspective";
    AntiprismSymmetryPerspective["__interfaces"] = ["com.vzome.core.editor.SymmetryPerspective"];


}
